spec/01-unit/000-new-dao/03-arguments_spec.lua

```diff
@@ -1,85 +1,146 @@
 local arguments    = require "kong.api.arguments"
+local Schema       = require "kong.db.schema"
 
+local infer_value = arguments.infer_value
+local infer       = arguments.infer
+local decode_arg  = arguments.decode_arg
+local decode      = arguments.decode
+local combine     = arguments.combine
 
-local decode_value = arguments.decode_value
-local decode       = arguments.decode
-local tonumber     = tonumber
-local null         = ngx.null
 
-
-describe("arguments.decode_value", function()
-  it("infers empty strings", function()
-    assert.equal(null, decode_value(""))
+describe("arguments.infer_value", function()
+  it("infers numbers", function()
+    assert.equal(2, infer_value("2", { type = "number" }))
+    assert.equal(2, infer_value("2", { type = "integer" }))
+    assert.equal(2.5, infer_value("2.5", { type = "number" }))
+    assert.equal(2.5, infer_value("2.5", { type = "integer" })) -- notice that integers are not rounded
   end)
 
   it("infers booleans", function()
-    assert.equal(true, decode_value("true"))
-    assert.equal(false, decode_value("false"))
+    assert.equal(false, infer_value("false", { type = "boolean" }))
+    assert.equal(true, infer_value("true", { type = "boolean" }))
   end)
 
-  it("infers numbers", function()
-    assert.equal(tonumber("123"), decode_value("123"))
-    assert.equal(tonumber("0.1"), decode_value("0.1"))
+  it("infers arrays and sets", function()
+    assert.same({ "a" }, infer_value("a",   { type = "array", elements = { type = "string" } }))
+    assert.same({ 2 },   infer_value("2",   { type = "array", elements = { type = "number" } }))
+    assert.same({ "a" }, infer_value({"a"}, { type = "array", elements = { type = "string" } }))
+    assert.same({ 2 },   infer_value({"2"}, { type = "array", elements = { type = "number" } }))
+
+    assert.same({ "a" }, infer_value("a",   { type = "set", elements = { type = "string" } }))
+    assert.same({ 2 },   infer_value("2",   { type = "set", elements = { type = "number" } }))
+    assert.same({ "a" }, infer_value({"a"}, { type = "set", elements = { type = "string" } }))
+    assert.same({ 2 },   infer_value({"2"}, { type = "set", elements = { type = "number" } }))
   end)
 
-  it("infers arrays", function()
-    assert.same(
-      { null, true, false, tonumber("123"), tonumber("0.1") },
-      decode_value { "", "true", "false", "123", "0.1" }
-    )
+  it("infers nulls from empty strings, arrays or sets", function()
+    assert.equal(ngx.null, infer_value("", { type = "string" }))
+    assert.equal(ngx.null, infer_value("", { type = "array" }))
+    assert.equal(ngx.null, infer_value("", { type = "set" }))
```

Comment:
If feels weird that `infer_value` returns null when it encounters "" for arrays and sets, but not for maps or records.

@hishamhm, what is your opinion on this?

Reply:
`""` for maps and records should infer `null` as well.